pgmpy/inference/Sampling.py

```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python3
+from itertools import product
+
+import networkx as nx
+from pandas import DataFrame
+
+from pgmpy.inference import Inference
+from pgmpy.models import BayesianModel
+from pgmpy.utils.mathext import sample_discrete
+from pgmpy.factors.Factor import State
+
+class BayesianModelSampling(Inference):
+    """
+    Class for sampling methods specific to Bayesian Models
+
+    Parameters
+    ----------
+    model: instance of BayesianModel
+        model on which inference queries will be computed
+
+
+    Public Methods
+    --------------
+    forward_sample(size)
+    """
+    def __init__(self, model):
+        if not isinstance(model, BayesianModel):
+            raise TypeError("model must an instance of BayesianModel")
+        Inference.__init__(self, model)
+        self.topological_order = nx.topological_sort(model)
+        self.cpds = {}
+        self.cpds = {node: model.get_cpds(node) for node in model.nodes()}
+
+    def forward_sample(self, size=1):
+        """
+        Generates sample(s) from joint distribution of the bayesian network.
+
+        Parameters
+        ----------
+        size: int
+            size of sample to be generated
+
+        Returns
+        -------
+        sampled: pandas.DataFrame
+            the generated samples
+
+        Examples
+        --------
+        >>> from pgmpy.models.BayesianModel import BayesianModel
+        >>> from pgmpy.factors.CPD import TabularCPD
+        >>> from pgmpy.inference.Sampling import BayesianModelSampling
+        >>> student = BayesianModel([('diff', 'grade'), ('intel', 'grade')])
+        >>> cpd_d = TabularCPD('diff', 2, [[0.6], [0.4]])
+        >>> cpd_i = TabularCPD('intel', 2, [[0.7], [0.3]])
+        >>> cpd_g = TabularCPD('grade', 3, [[0.3, 0.05, 0.9, 0.5], [0.4, 0.25,
+        ...                0.08, 0.3], [0.3, 0.7, 0.02, 0.2]],
+        ...                ['intel', 'diff'], [2, 2])
+        >>> student.add_cpds(cpd_d, cpd_i, cpd_g)
+        >>> inference = BayesianModelSampling(student)
+        >>> inference.forward_sample(2)
+                diff       intel       grade
+        0  (diff, 1)  (intel, 0)  (grade, 1)
+        1  (diff, 1)  (intel, 0)  (grade, 2)
+        """
+        sampled = DataFrame(index=range(size), columns=self.topological_order)
+        for node in self.topological_order:
+            cpd = self.cpds[node]
+            if cpd.evidence:
+                # memoizing calls to cpd.reduce
+                weights = {}
+                card_iter = [range(card) for card in cpd.evidence_card]
+                for tup in product(*card_iter):
+                    evid = [State(v, s) for v, s in zip(cpd.evidence, tup)]
+                    weights[tup] = cpd.reduce(evid, inplace=False).values
+                # can now fill values, row by row
+                for i in range(size):
```

Comment:
@pratyakshs for loops are costly. So, what you can do here is that you can create a 2-D array of states and a 2-D array for weights as:
For states:
[tuple/array of states of evidence vars for 1st sample,
 tuple/array of states of evidence vars for 2nd sample,
 ... ]

and similarly for weights:
[weight arr for 1st sample,
 weight arr for 2nd sample,
 ...]

and modify `sample_discrete` accordingly so that it computes all the values in a single vector operation.